src/components/chart/down-sample.ts

```diff
@@ -0,0 +1,67 @@
+import type { LineSeriesOption } from "echarts";
+
+export function downSampleLineData(
+  data: LineSeriesOption["data"],
+  chartWidth: number,
+  minX?: number,
+  maxX?: number
+) {
+  if (!data || data.length < 10) {
+    return data;
+  }
+  const width = chartWidth * window.devicePixelRatio;
+  if (data.length <= width) {
+    return data;
+  }
+  const min = minX ?? data[0]![0];
+  const max = maxX ?? data[data.length - 1]![0];
+  const step = Math.floor((max - min) / width);
```

Comment:
Since we add a min and max for every frame we could end up with a result that has the exact same number of points as the data, if the data is spread out perfect over the graph and has 2 times the data points of width right?

Reply:
Yes, this is how echarts does it too.